src/library/Query.js

```diff
@@ -33,11 +53,42 @@ function query(httpMethod, deployment, apiKey, qry, payload) {
 }
 
 function limitQuery(httpMethod, deployment, apiKey, queryPath, limit, offset, payload, address) {
-  // validate and normalize limit and offset
-  const queryOptions = buildQueryOptions(limit, offset, address);
+  const separationLimit = 30;
+  const limitChecked = checkLimit(limit);
+  const offsetChecked = checkOffset(offset);
+  const results = {
+    status: null,
+    message: null,
+  };
+  // MBQUERY, MBCUSTOMQUERY has rows in result but MBEVENTS has not
+  const hasRows = !/^events$/.test(queryPath);
+  const rows = [];
+  let queryRows = [];
+
+  let offsetNext = offsetChecked;
+  do {
+    // validate and normalize limit and offset
+    const queryOptions = buildQueryOptions(separationLimit, offsetNext, address);
+    offsetNext += separationLimit;
+
+    // query
+    const queryResult = query(httpMethod, deployment, apiKey, queryPath + queryOptions, payload);
+    queryRows = hasRows ? queryResult.result.rows : queryResult.result;
+    const overflowedLength = (rows.length + queryRows.length) - limitChecked;
+    if (overflowedLength > 0) {
+      // either add the entire array of "queryRows" to "rows", or only up to the total limit
+      // works because if "end" is greater than the length of the array, it only extracts
+      // up to the end of the array
+      rows.push(...queryRows.slice(0, queryRows.length - overflowedLength));
+    } else {
+      rows.push(...queryRows);
+    }
+
+    results.status = queryResult.status;
+    results.message = queryResult.message;
+  } while (queryRows.length > 0 && rows.length < limitChecked);
```

Comment:
We should also check `... && queryRows.length < separationLimit` to handle the case where we receive some results.

In this case, there is no need for the second query to `.../queries/.../results`:

![image](https://user-images.githubusercontent.com/5861379/101886480-e4a92180-3bde-11eb-9fd5-0fe6a93d9d37.png)

Reply:
That makes sense!

Reply:
Do you mean `&& limitChecked != separationLimit` not `&& queryRows.length < separationLimit`?

if limitChecked and separationLimit are same then just one time API call happens.

I think in the case of adding `&& queryRows.length < separationLimit`, it will also call API one more time if the number of results is smaller than `separationLimit`.

What do you think?

Reply:
Yes! Good call, I think you are correct.